Implement federated exemplar API

goku321 opened the pull request:
<!--
    Keep PR title verbose enough and add prefix telling
    about what components it touches e.g "query:" or ".*:"
-->
Signed-off-by: Deepak sah.sslpu@gmail.com

<!--
    Don't forget about CHANGELOG!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Thanos <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR such as https://github.com/thanos-io/thanos/pull/<PR-id>
    <Component> Component affected by your changes such as Query, Store, Receive.
-->

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

Fixes: #3435 

This PR follows the PR #3350 and comment #3345 to implement the Prometheus exemplar API

## Changes

<!-- Enumerate changes you made -->
Add federated exemplar api to query and sidecar.

## Verification

testing in progress

goku321 commented:
@bwplotka @yeya24 I've tried to complete this as much as possible but there are still few gaps. A review would really be helpful to find out all the issues this pr has.

Changes I'm not sure about:
1. [pkg/exemplars/exemplarspb/custom.go](https://github.com/thanos-io/thanos/pull/3846/files#diff-0d77e7c7241aa0d036e0050643e9b9d5cfd0d35180614e101271ee169ad644b0)
2. [pkg/exemplars/exemplars.go](https://github.com/thanos-io/thanos/pull/3846/files#diff-d09c4014e056aee5c67a76aca788de273aecb440edefc8a3b040bc77e27f2c38)
3. [test/e2e/exemplars_api_test.go](https://github.com/thanos-io/thanos/pull/3846/files?file-filters%5B%5D=.go#diff-9c0f406d218e2d082144a299439ce68d6ae705fb1426380eaf4eb615840d56a4)

Thank you for your support and guidance.

yeya24 commented:
Thanks for the great work! Looks perfect at the first glance. I will do some detailed review next week.

Have you tested this manually with the WIP upstream branch? https://github.com/prometheus/prometheus/pull/6635

goku321 commented:
>Have you tested this manually with the WIP upstream branch? prometheus/prometheus#6635

Not yet 😅 . Will be doing it very soon.

yeya24 commented:
I fixed some existing problems in this pr and it is working now! Will add some detailed comments tomorrow. Feel free to check my commit [here](https://github.com/yeya24/thanos/commit/1b6abfcb35e2a7153a9dabda921926a7f72b8c7f).

If you want to experiment with this, feel free to try this tns app https://github.com/yeya24/tns/tree/thanos/production/docker-compose. I added Thanos sidecar and querier there.

![Screenshot from 2021-02-27 01-34-54](https://user-images.githubusercontent.com/25150124/109377875-05a71780-789c-11eb-9004-9a342d986d13.png)

It works smoothly. Thanks @goku321!

yeya24 commented on pkg/api/query/v1.go:
```diff
@@ -152,6 +158,8 @@ func (qapi *QueryAPI) Register(r *route.Router, tracer opentracing.Tracer, logge
 
 	r.Get("/metadata", instr("metadata", NewMetricMetadataHandler(qapi.metadatas, qapi.enableMetricMetadataPartialResponse)))
+
+	r.Get("/exemplars", instr("exemplars", NewExemplarsHandler(qapi.exemplars, qapi.enableExemplarPartialResponse)))
```
Let's be consistent with the upstream PR https://github.com/prometheus/prometheus/pull/6635/files#diff-9bcb876da5d17ef1046ab5ea531bbaaf64d753a7beaa2cb11a3fe6dd94e16563R349.

The endpoint is `query_exemplars`

yeya24 commented on pkg/api/query/v1.go:
```diff
@@ -679,6 +687,36 @@ func NewRulesHandler(client rules.UnaryClient, enablePartialResponse bool) func(
+	}
+
+	return func(r *http.Request) (interface{}, []error, *api.ApiError) {
+		typeParam := r.URL.Query().Get("type")
```
What's this param for?
Let's be consistent with the upstream pr.
The params we want to have are:

```
- `query=<string>`: Prometheus expression query string.
- `start=<rfc3339 | unix_timestamp>`: Start timestamp.
- `end=<rfc3339 | unix_timestamp>`: End timestamp.
```

yeya24 commented on pkg/exemplars/exemplarspb/rpc.proto:
```diff
@@ -0,0 +1,73 @@
+    string instance = 1 [(gogoproto.jsontag) = "instance"];
+    string job = 2 [(gogoproto.jsontag) = "job"];
+    string service = 3 [(gogoproto.jsontag) = "service"];
+}
```
This seems not used. Let's remove it.

yeya24 commented on pkg/exemplars/exemplarspb/rpc.proto:
```diff
@@ -0,0 +1,73 @@
+    double value = 2 [(gogoproto.jsontag) = "value"];
+    google.protobuf.Timestamp ts = 3 [(gogoproto.jsontag) = "timestamp", (gogoproto.stdtime) = true, (gogoproto.nullable) = false ];
+    bool hasts = 4 [(gogoproto.jsontag) = "hasTimestamp"];
+}
```
Please add a new line.

yeya24 commented on pkg/exemplars/exemplarspb/rpc.proto:
```diff
@@ -0,0 +1,73 @@
+message Exemplar {
+    ZLabelSet labels = 1 [(gogoproto.jsontag) = "labels", (gogoproto.nullable) = false];
+    double value = 2 [(gogoproto.jsontag) = "value"];
+    google.protobuf.Timestamp ts = 3 [(gogoproto.jsontag) = "timestamp", (gogoproto.stdtime) = true, (gogoproto.nullable) = false ];
```
Let's use int64 type here to be consistent with the TS type in prometheus https://github.com/prometheus/prometheus/pull/6635/files#diff-fba4b55f88ed50da9d6d2624ba745f9347cdc87a80bac431a090aee67a49c02fR23

yeya24 commented on pkg/promclient/promclient.go:
```diff
@@ -758,3 +759,30 @@ func (c *Client) MetadataInGRPC(ctx context.Context, base *url.URL, metric strin
+// NOTE: This method is tested in pkg/store/prometheus_test.go against Prometheus.
+func (c *Client) ExemplarsInGRPC(ctx context.Context, base *url.URL, typeExemplars string) ([]*exemplarspb.ExemplarData, error) {
+	u := *base
+	u.Path = path.Join(u.Path, "/api/v1/exemplars")
```
Ditto. Let's change the endpoint.

yeya24 commented on pkg/promclient/promclient.go:
```diff
@@ -758,3 +759,30 @@ func (c *Client) MetadataInGRPC(ctx context.Context, base *url.URL, metric strin
+	// Prometheus does not support PartialResponseStrategy, and probably would never do. Make it Abort by default.
+	for _, g := range m.Data {
+		g.PartialResponseStrategy = storepb.PartialResponseStrategy_ABORT
+	}
```
I guess we don't need to keep this `PartialResponseStrategy` field in the proto. Let's remove it for now.

yeya24 commented on pkg/exemplars/exemplarspb/rpc.proto:
```diff
@@ -0,0 +1,73 @@
+// message Label {
+//     string name = 1 [(gogoproto.jsontag) = "name"];
+//     string value = 2 [(gogoproto.jsontag) = "value"];
+// }
```
Let's remove this

yeya24 commented on pkg/exemplars/exemplars.go:
```diff
@@ -0,0 +1,145 @@
+
+	warnings []error
+	data     []*exemplarspb.ExemplarData
+}
```
Let's add two additional methods for it https://github.com/yeya24/thanos/commit/1b6abfcb35e2a7153a9dabda921926a7f72b8c7f#diff-d09c4014e056aee5c67a76aca788de273aecb440edefc8a3b040bc77e27f2c38R39-R55, otherwise it will panic.

yeya24 commented:
@goku321 Let me know your availability. If it is acceptable I can open a pr to your branch as I have a working version already https://github.com/yeya24/thanos/commit/1b6abfcb35e2a7153a9dabda921926a7f72b8c7f.

goku321 commented:
Hi @yeya24 , I was a bit caught up, thanks for your patience.

By all means. Please go ahead with the pr. I am glad to have your help on this. I really appreciate your time and effort on fixing this pr.
>If it is acceptable I can open a pr to your branch as I have a working version already

yeya24 commented:
It is not urgent. Anyway, we need to wait for the upstream Prometheus pr first XD

goku321 commented:
> I fixed some existing problems in this pr and it is working now! Will add some detailed comments tomorrow. Feel free to check my commit [here](https://github.com/yeya24/thanos/commit/1b6abfcb35e2a7153a9dabda921926a7f72b8c7f).
> 
> If you want to experiment with this, feel free to try this tns app https://github.com/yeya24/tns/tree/thanos/production/docker-compose. I added Thanos sidecar and querier there.
> 
> ![Screenshot from 2021-02-27 01-34-54](https://user-images.githubusercontent.com/25150124/109377875-05a71780-789c-11eb-9004-9a342d986d13.png)
> 
> It works smoothly. Thanks @goku321!

Sure, I'll try out the tns app to get a better understanding.

squat commented on pkg/exemplars/proxy.go:
```diff
@@ -0,0 +1,139 @@
+		return err
+	}
+
+	for _, e := range exemplars {
```
It looks like the proxy is not deduplicating at all. Should we deduplicate once we've collected all of the responses?

squat commented on pkg/exemplars/proxy.go:
```diff
@@ -0,0 +1,139 @@
+)
+
+// Proxy implements exemplarspb.Exemplars gRPC that fanouts requests to
+// given exemplarspb.Exemplars and de-duplication on the way.
```
Let's update this comment. Currently, Proxy itself does not deduplicate. Instead, the it seems to be a responsibility of the caller of proxy.Exemplars to deduplication. Alternatively, we can make it so that the Proxy needs to deduplication.

yeya24 commented on pkg/exemplars/exemplarspb/rpc.proto:
```diff
@@ -0,0 +1,55 @@
+    ZLabelSet labels = 1 [(gogoproto.jsontag) = "labels", (gogoproto.nullable) = false];
+    double value = 2 [(gogoproto.jsontag) = "value"];
+    int64 ts = 3 [(gogoproto.jsontag) = "timestamp"];
+    bool hasts = 4 [(gogoproto.jsontag) = "hasTimestamp"];
```
Let's remove this field? Since it is just an internal implementation in Prometheus and all exemplars we get from Prometheus will have a timestamp

yeya24 commented on test/e2e/e2ethanos/services.go:
```diff
@@ -75,6 +80,7 @@ func NewPrometheus(sharedDir string, name string, config, promImage string) (*e2
 			"--storage.tsdb.max-block-duration": "2h",
 			"--log.level":                       infoLogLevel,
 			"--web.listen-address":              ":9090",
+			"--enable-feature":                  strings.Join(enableFeatures, ","),
```
This flag should be added only if `len(enableFeatures) > 0`. Let's update it and see if E2E tests work or not

yeya24 commented on test/e2e/metadata_api_test.go:
```diff
@@ -50,6 +50,7 @@ func TestMetadataAPI_Fanout(t *testing.T) {
 		nil,
 		nil,
+		nil,
 		[]string{sidecar1.GRPCNetworkEndpoint(), sidecar2.GRPCNetworkEndpoint()},
```
The order needs to be changed. In this test case, metadata endpoints should be `[]string{sidecar1.GRPCNetworkEndpoint(), sidecar2.GRPCNetworkEndpoint()}` and exemplar endpoints should be nil.

yeya24 commented on test/e2e/e2ethanos/services.go:
```diff
@@ -38,7 +43,7 @@ var defaultBackoffConfig = util.BackoffConfig{
 // TODO(bwplotka): Run against multiple?
 func DefaultPrometheusImage() string {
-	return "quay.io/prometheus/prometheus:v2.19.3"
+	return "quay.io/prometheus/prometheus:main"
```
Let's change the image back to `v2.19.3`.
Updating the image to the latest main causes our Rule E2E test broken. Same as https://github.com/thanos-io/thanos/pull/3905#discussion_r591791448.

As I think it is okay to add the E2E test in a later pr, we don't need to update the image now.

yeya24 commented on test/e2e/e2ethanos/services.go:
```diff
@@ -66,16 +71,21 @@ func NewPrometheus(sharedDir string, name string, config, promImage string) (*e2
+	})
+	if len(enableFeatures) > 0 {
+		args = append(args, "--enable-feature="+strings.Join(enableFeatures, ","))
+	}
```
This should be removed as we rollback the Prometheus image.

yeya24 commented on test/e2e/e2ethanos/services.go:
```diff
@@ -55,7 +60,7 @@ func DefaultImage() string {
 }
 
-func NewPrometheus(sharedDir string, name string, config, promImage string) (*e2e.HTTPService, string, error) {
+func NewPrometheus(sharedDir string, name string, config, promImage string, enableFeatures ...string) (*e2e.HTTPService, string, error) {
```
Let's remove the last param

yeya24 commented on test/e2e/e2ethanos/services.go:
```diff
@@ -85,8 +95,8 @@ func NewPrometheus(sharedDir string, name string, config, promImage string) (*e2
-func NewPrometheusWithSidecar(sharedDir string, netName string, name string, config, promImage string) (*e2e.HTTPService, *Service, error) {
-	prom, dataDir, err := NewPrometheus(sharedDir, name, config, promImage)
+func NewPrometheusWithSidecar(sharedDir string, netName string, name string, config, promImage string, enableFeatures ...string) (*e2e.HTTPService, *Service, error) {
+	prom, dataDir, err := NewPrometheus(sharedDir, name, config, promImage, enableFeatures...)
```
Let's remove the last param

yeya24 commented on test/e2e/e2ethanos/services.go:
```diff
@@ -29,6 +30,10 @@ import (
 
+const (
+	ExemplarStorage = "exemplar-storage"
+)
```
Let's remove this

yeya24 commented on pkg/exemplars/exemplars_test.go:
```diff
@@ -0,0 +1,66 @@
+						{Name: "service", Value: "bar"},
+					}},
+				},
+			},
```
Let's add more test cases here:
1. There are one or more exemplars in the input data
2. It can contain more than one series

yeya24 commented on pkg/exemplars/exemplarspb/custom.go:
```diff
@@ -0,0 +1,105 @@
+	if d := labels.Compare(s1.SeriesLabels.PromLabels(), s2.SeriesLabels.PromLabels()); d != 0 {
+		return d
+	}
+	return 0
```
We can just use `labels.Compare` here, no need to use `if d := xxx; d != 0` pattern since we removed the `HasTS` field already.

yeya24 commented on pkg/exemplars/exemplarspb/custom.go:
```diff
@@ -0,0 +1,105 @@
+		return d
+	}
+
+	return 0
```
Same here. Return `big.NewFloat(e1.Value).Cmp(big.NewFloat(e2.Value))` is fine.

yeya24 reviewed (approved):
@goku321 Thanks for the great work! This is super close now, only some nits and testcases are left.

yeya24 commented on pkg/api/query/v1.go:
```diff
@@ -683,6 +692,35 @@ func NewRulesHandler(client rules.UnaryClient, enablePartialResponse bool) func(
+			return nil, nil, &api.ApiError{Typ: api.ErrorBadData, Err: err}
+		}
+
+		req := &exemplarspb.ExemplarsRequest{
```
Looks like some of my original changes are missing in your rebase. Can you please check https://github.com/yeya24/thanos/commit/1b6abfcb35e2a7153a9dabda921926a7f72b8c7f#diff-f14ed7ae9f9be7d76dea77aa31304b22de00af7e5f08720a7b6e75f721caf24aR700-R711?

goku321 replied:
Oh, I missed this one. Thanks for pointing it out.

yeya24 commented on pkg/exemplars/exemplarspb/custom.go:
```diff
@@ -0,0 +1,105 @@
+	}{
+		Labels:    labelpb.ZLabelsToPromLabels(m.Labels.Labels),
+		TimeStamp: model.Time(m.Ts),
+		Value:     model.SampleValue(m.Ts),
```
This should be `m.Value`

yeya24 reviewed (approved):
Perfect work @goku321. One last thing, can you please update the changelog?
I think this pr is ready to merge. But it would be better to have another :eyes: for a final look.

Note: There is no E2E API test for this feature now as we don't have an app instrumented with exemplars. We can move E2E part to another pr.

squat reviewed (approved):
Looks clean to me!
